aten/src/ATen/native/mps/operations/Copy.mm

```diff
@@ -113,21 +113,28 @@ void copy_cast_mps(at::Tensor& dst, const at::Tensor& src,
     src = src_;
   }
   id<MTLBuffer> sourceBuffer = getMTLBufferStorage(src);
-  const size_t src_size = src.nbytes();
-  // if there's anything wrong with source, we shouldn't return dst_ silently and must error out.
-  TORCH_CHECK(sourceBuffer && src_size > 0);
+  size_t src_total_size = src_.is_view() ? at::detail::computeStorageNbytesContiguous(src.sizes(), src.element_size(), src.storage_offset()) :
+                                           src.nbytes();
+  size_t size_to_copy = src.nbytes();
 
   // In case of dtype change, first convert src inplace
   if (src_.dtype() != dst_.dtype()) {
     copy_cast_mps(dst, src, sourceBuffer, sourceBuffer);
+    // Use the element size of dst to calculate the total size after casting
+    size_to_copy = (size_to_copy / src.element_size()) * dst.element_size();
   }
 
+  // if there's anything wrong with source, we shouldn't return dst_ silently and must error out.
+  TORCH_CHECK(sourceBuffer && size_to_copy > 0);
```

Comment:
Shouldn't these TORCH_CHECK()s be replaced with TORCH_INTERNAL_ASSERT()? Because it's not user's error if the assert trigger. It's our internal error at that point and the user should be instructed to file a bug (the TORCH_INTERNAL_ASSERT() does that).

Reply:
That's fair, @DenisVieriu97 , lets change that.